erigon-lib/state/inverted_index.go

```diff
@@ -1476,42 +1478,115 @@ func (iit *InvertedIndexRoTx) IterateChangedKeys(startTxNum, endTxNum uint64, ro
 }
 
 // collate [stepFrom, stepTo)
-func (ii *InvertedIndex) collate(ctx context.Context, step uint64, roTx kv.Tx) (map[string]*roaring64.Bitmap, error) {
+func (ii *InvertedIndex) collate(ctx context.Context, step uint64, roTx kv.Tx) (InvertedIndexCollation, error) {
 	stepTo := step + 1
 	txFrom, txTo := step*ii.aggregationStep, stepTo*ii.aggregationStep
 	start := time.Now()
-	defer mxCollateTook.ObserveDuration(start)
+	defer mxCollateTookIndex.ObserveDuration(start)
 
 	keysCursor, err := roTx.CursorDupSort(ii.indexKeysTable)
 	if err != nil {
-		return nil, fmt.Errorf("create %s keys cursor: %w", ii.filenameBase, err)
+		return InvertedIndexCollation{}, fmt.Errorf("create %s keys cursor: %w", ii.filenameBase, err)
 	}
 	defer keysCursor.Close()
-	indexBitmaps := map[string]*roaring64.Bitmap{}
+
+	collector := etl.NewCollector(ii.indexKeysTable, ii.iiCfg.dirs.Tmp, etl.NewSortableBuffer(CollateETLRAM), ii.logger)
+	defer collector.Close()
+
 	var txKey [8]byte
 	binary.BigEndian.PutUint64(txKey[:], txFrom)
+
 	for k, v, err := keysCursor.Seek(txKey[:]); k != nil; k, v, err = keysCursor.Next() {
 		if err != nil {
-			return nil, fmt.Errorf("iterate over %s keys cursor: %w", ii.filenameBase, err)
+			return InvertedIndexCollation{}, fmt.Errorf("iterate over %s keys cursor: %w", ii.filenameBase, err)
 		}
 		txNum := binary.BigEndian.Uint64(k)
 		if txNum >= txTo { // [txFrom; txTo)
 			break
 		}
-		bitmap, ok := indexBitmaps[string(v)]
-		if !ok {
-			bitmap = bitmapdb.NewBitmap64()
-			indexBitmaps[string(v)] = bitmap
+		if err := collector.Collect(v, k); err != nil {
+			return InvertedIndexCollation{}, fmt.Errorf("collect %s history key [%x]=>txn %d [%x]: %w", ii.filenameBase, k, txNum, k, err)
 		}
-		bitmap.Add(txNum)
-
 		select {
 		case <-ctx.Done():
-			return nil, ctx.Err()
+			return InvertedIndexCollation{}, ctx.Err()
 		default:
 		}
 	}
-	return indexBitmaps, nil
+
+	var (
+		coll = InvertedIndexCollation{
+			iiPath: ii.efFilePath(step, stepTo),
+		}
+		closeComp bool
+	)
+	defer func() {
+		if closeComp {
+			coll.Close()
+		}
+	}()
+
+	comp, err := seg.NewCompressor(ctx, "snapshots", coll.iiPath, ii.dirs.Tmp, seg.MinPatternScore, ii.compressWorkers, log.LvlTrace, ii.logger)
+	if err != nil {
+		return InvertedIndexCollation{}, fmt.Errorf("create %s compressor: %w", ii.filenameBase, err)
+	}
+	coll.writer = NewArchiveWriter(comp, ii.compression)
+
+	var (
+		prevEf      []byte
+		prevKey     []byte
+		initialized atomic.Bool
+		bitmap      = bitmapdb.NewBitmap64()
```

Comment:
maybe need add `bitmapdb.ReturnToPool(bitmap)`

Reply:
Doesnt bitmap.Reset() covers the issue? because it started to fill with new txNums straight after

Reply:
added this call both to history and ii defers.